File: 2017/09/rs/src/main.rs
const GROUP_START: char = '{';
const GROUP_END: char = '}';
const GARBAGE_START: char = '<';
const GARBAGE_END: char = '>';
const ESCAPE: char = '!';

fn solve(stream: &str) -> (u32, u32) {
    let mut group_score = 0;
    let mut garbage_count = 0;
    let mut depth = 0;
    let mut in_garbage = false;
    let mut escape = false;
    for c in stream.chars() {
        if escape {
            escape = false;
        } else if in_garbage {
            if c == ESCAPE {
                escape = true;
            } else if c == GARBAGE_END {
                in_garbage = false;
            } else {
                garbage_count += 1;
            }
        } else if c == GARBAGE_START {
            in_garbage = true;
        } else if c == GROUP_START {
            depth += 1;
        } else if c == GROUP_END {
            group_score += depth;
            depth -= 1;
        }
    }
    (group_score, garbage_count)
}

fn get_input(file_path: &String) -> String {
    String::from(std::fs::read_to_string(file_path).expect("Error reading input file!"))
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = std::time::Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    let end = now.elapsed().as_secs_f32();
    println!("P1: {}", part1_result);
    println!("P1: {}", part2_result);
    println!();
    println!("Time: {:.7}", end);
}

File: 2017/09/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>

typedef char *Input;
typedef struct
{
    int part1;
    int part2;
} Results;

#define GROUP_START '{'
#define GROUP_END '}'
#define GARBAGE_START '<'
#define GARBAGE_END '>'
#define ESCAPE '!'

Results solve(Input input)
{
    int groupScore = 0, garbageCount = 0, depth = 0, inGarbage = 0, escape = 0;
    char c;
    while (c = *(input++))
    {
        if (escape)
            escape = 0;
        else if (inGarbage)
        {
            if (c == ESCAPE)
                escape = 1;
            else if (c == GARBAGE_END)
                inGarbage = 0;
            else
                garbageCount++;
        }
        else if (c == GARBAGE_START)
            inGarbage = 1;
        else if (c == GROUP_START)
            depth++;
        else if (c == GROUP_END)
            groupScore += depth--;
    }
    return (Results){groupScore, garbageCount};
}

Input getInput(char *filePath)
{
    FILE *file;
    if ((file = fopen(filePath, "r")) == NULL)
    {
        perror("Error reading input file!\n");
        exit(1);
    }
    fseek(file, 0, SEEK_END);
    long length = ftell(file);
    rewind(file);
    char *content = malloc(length);
    fread(content, 1, length, file);
    fclose(file);
    return content;
}

void freeInput(Input input)
{
    free(input);
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        perror("Please, add input file path as parameter");
        exit(1);
    }
    struct timeval starts, ends;
    gettimeofday(&starts, NULL);
    Input input = getInput(argv[1]);
    Results results = solve(input);
    gettimeofday(&ends, NULL);
    freeInput(input);
    printf("P1: %d\n", results.part1);
    printf("P2: %d\n\n", results.part2);
    printf("Time: %.7f\n", (double)((ends.tv_sec - starts.tv_sec) * 1000000 + ends.tv_usec - starts.tv_usec) / 1000000);
    return 0;
}
